hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/datetime/HazelcastExtractFunction.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite.validate.operators.datetime;
+
+import com.hazelcast.sql.impl.calcite.validate.HazelcastCallBinding;
+import com.hazelcast.sql.impl.calcite.validate.operand.AnyOperandChecker;
+import com.hazelcast.sql.impl.calcite.validate.operand.OperandCheckerProgram;
+import com.hazelcast.sql.impl.calcite.validate.operand.TypedOperandChecker;
+import com.hazelcast.sql.impl.calcite.validate.operators.ReplaceUnknownOperandTypeInference;
+import com.hazelcast.sql.impl.calcite.validate.operators.common.HazelcastFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+public final class HazelcastExtractFunction extends HazelcastFunction {
+    public static final HazelcastExtractFunction INSTANCE = new HazelcastExtractFunction();
+
+    private HazelcastExtractFunction() {
+        super(
+                "EXTRACT",
+                SqlKind.EXTRACT,
+                ReturnTypes.DOUBLE_NULLABLE,
```

Comment:
EXTRACT function, according to the standard, is supposed to return exact integer type, except for SECOND, where it's supposed to return DECIMAL, not DOUBLE.

Reply:
There is a Postgres divergence from the standard here. 
Postgres specifies `EXTRACT` returns double precision and I decided to stick with it.

However, I agree that fixed point numeric is better fit here since the `EXTRACT(SECOND FROM ...)` will return a value with at most 6-digits precision.

I will change the return type, yet I wonder what is the rationale behind Postgres' decision of double precision return type.

Reply:
I don't have a strong opinion in this particular case. It makes sense that since timestamps are exact, the extracted parts should be too. But DOUBLE is precise for integers up to 52 bits if I'm not mistaken, so it's fine to store epoch milliseconds for some 140k years, so I agree with following PG. BigDecimal is impractical here. The only difference can be in fractional seconds, but one can also extract MICROSECONDS, though not NANOSECONDS...
I'm marking this as resolved.